Sutter stage updates

zacsimile opened the pull request:
Changes to Sutter code, as suggested by Bruce at ASI.  Flush buffers more often, read byte-by-byte. We should keep these changes as they are theoretically safer than the previous code, but initial testing suggests no decrease in frequency of errors :/.

AdvancedImagingUTSW committed:
Flush buffers more often, read byte-by-byte

Co-Authored-By: Zach Marin <zacsimile@users.noreply.github.com>

AdvancedImagingUTSW commented:
The only other thing I can think of is that this error is unique to our code, and was not experienced with the LabView code. If you want to dig into that, you could see how Gerard did it.

zacsimile commented:
Thanks! I'll have a look at the LabView.

AdvancedImagingUTSW committed:
Update stage_sutter.py

I implemented some basic logic into the MP285 to make sure that we do not tell the stage to move, unless the move is greater than some threshold. The threshold chosen arbitrarily was 200 nm. It is specific to this class, and the MP285 should never be used to acquire image slices for a z-stack, which is the only time we would have a step that size.

Side-effect: Much faster since we are not pinging the stage at every z-slice, and this could help us avoid the stupid scrambled screen if it arises from too many pings.

AdvancedImagingUTSW committed:
Fix for tiff image_writer

Comparison was a bits bytes mismatch. File size estimate for tiff now matches what is saved. shutil measurement of OS hard drive space is correct.

Did not implement for other file types yet. Quickly tried to, but ran into a random error with HDF5, and then ran out of time as kids started waking up.

Rather than throwing a user warning, we just print it now.

AdvancedImagingUTSW committed:
Don't pre-calculate z-stack waveform for Galvo stages

Co-Authored-By: Zach Marin <zacsimile@users.noreply.github.com>
Co-Authored-By: Annie Wang <6161065+annie-xd-wang@users.noreply.github.com>

annie-xd-wang commented on src/aslm/model/devices/APIs/sutter/MP285.py:
```diff
@@ -185,17 +185,18 @@ def get_current_position(self):
         # print(f"sending: {command}")
         position_information = b""
         for _ in range(self.n_waits):
```
The reading time should be at least 13 since we need to read out 13 bytes and read one byte each time. However, self.n_waits means how many times we should try to ping the machine during the specified waiting time.

zacsimile commented:
Strange issue...

```
2023-07-25 14:21:24,879 - model - DEBUG - stage_sutter: MP-285 - Position: {'y_pos': -11970.6, 'x_pos': 11715.4, 'f_pos': 1578.88}
2023-07-25 14:21:38,098 - model - DEBUG - stage_sutter: MP-285 - Position: {'y_pos': -11970.6, 'x_pos': 11715.4, 'f_pos': 1578.88}
2023-07-25 14:21:38,597 - model - DEBUG - feature_container: running signal node: ZStackAcquisition
2023-07-25 14:21:38,609 - model - DEBUG - filter_wheel_sutter: SutterFilterWheel - Moving to Position 0
2023-07-25 14:21:38,759 - model - DEBUG - feature_container: SignalContainer - Traceback (most recent call last):
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\features\feature_container.py", line 208, in run
    result, is_end = self.curr_node.run(*args, wait_response=wait_response)
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\features\feature_container.py", line 91, in run
    self.node_funcs["init"]()
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\features\common_features.py", line 315, in pre_signal_func
    self.model.active_microscope.prepare_next_channel()
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\microscope.py", line 527, in prepare_next_channel
    self.daq.prepare_acquisition(channel_key, self.current_exposure_time)
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\devices\daq\daq_ni.py", line 397, in prepare_acquisition
    self.create_analog_output_tasks(channel_key)
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\devices\daq\daq_ni.py", line 365, in create_analog_output_tasks
    [
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\devices\daq\daq_ni.py", line 366, in <listcomp>
    v["waveform"][channel_key][:max_sample]
KeyError: 'channel_1'

2023-07-25 14:21:39,015 - model - INFO - model: ASLM Model - Running data process, get frames [0]
2023-07-25 14:21:39,050 - model - INFO - model: ASLM Model - Sent through pipe0
2023-07-25 14:21:39,551 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:21:39,551 - model - INFO - model: ASLM Model - Waiting 10
2023-07-25 14:21:40,052 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:21:40,052 - model - INFO - model: ASLM Model - Waiting 9
2023-07-25 14:21:40,554 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:21:40,554 - model - INFO - model: ASLM Model - Waiting 8
2023-07-25 14:21:41,055 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:21:41,055 - model - INFO - model: ASLM Model - Waiting 7
2023-07-25 14:21:41,557 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:21:41,557 - model - INFO - model: ASLM Model - Waiting 6
2023-07-25 14:21:42,057 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:21:42,057 - model - INFO - model: ASLM Model - Waiting 5
2023-07-25 14:21:42,559 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:21:42,559 - model - INFO - model: ASLM Model - Waiting 4
2023-07-25 14:21:43,059 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:21:43,059 - model - INFO - model: ASLM Model - Waiting 3
2023-07-25 14:21:43,561 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:21:43,561 - model - INFO - model: ASLM Model - Waiting 2
2023-07-25 14:21:44,063 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:21:44,064 - model - INFO - model: ASLM Model - Waiting 1
2023-07-25 14:21:44,064 - model - INFO - model: ASLM Model - Data thread stopped.
2023-07-25 14:21:44,064 - model - INFO - model: ASLM Model - Received frames in total: 1
2023-07-25 14:27:19,394 - model - DEBUG - stage_sutter: MP-285 - Position: {'y_pos': -11970.6, 'x_pos': 11715.4, 'f_pos': 1578.88}
2023-07-25 14:27:19,710 - model - DEBUG - feature_container: running signal node: ZStackAcquisition
2023-07-25 14:27:19,720 - model - DEBUG - filter_wheel_sutter: SutterFilterWheel - Moving to Position 0
2023-07-25 14:27:19,743 - model - DEBUG - feature_container: SignalContainer - Traceback (most recent call last):
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\features\feature_container.py", line 208, in run
    result, is_end = self.curr_node.run(*args, wait_response=wait_response)
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\features\feature_container.py", line 91, in run
    self.node_funcs["init"]()
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\features\common_features.py", line 315, in pre_signal_func
    self.model.active_microscope.prepare_next_channel()
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\microscope.py", line 527, in prepare_next_channel
    self.daq.prepare_acquisition(channel_key, self.current_exposure_time)
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\devices\daq\daq_ni.py", line 397, in prepare_acquisition
    self.create_analog_output_tasks(channel_key)
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\devices\daq\daq_ni.py", line 365, in create_analog_output_tasks
    [
  File "c:\users\spectral\documents\github\aslm\src\aslm\model\devices\daq\daq_ni.py", line 366, in <listcomp>
    v["waveform"][channel_key][:max_sample]
KeyError: 'channel_1'

2023-07-25 14:27:20,000 - model - INFO - model: ASLM Model - Running data process, get frames [0]
2023-07-25 14:27:20,000 - model - INFO - model: ASLM Model - Sent through pipe0
2023-07-25 14:27:20,501 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:27:20,501 - model - INFO - model: ASLM Model - Waiting 10
2023-07-25 14:27:21,002 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:27:21,002 - model - INFO - model: ASLM Model - Waiting 9
2023-07-25 14:27:21,502 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:27:21,502 - model - INFO - model: ASLM Model - Waiting 8
2023-07-25 14:27:22,004 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:27:22,004 - model - INFO - model: ASLM Model - Waiting 7
2023-07-25 14:27:22,506 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:27:22,506 - model - INFO - model: ASLM Model - Waiting 6
2023-07-25 14:27:23,008 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:27:23,008 - model - INFO - model: ASLM Model - Waiting 5
2023-07-25 14:27:23,509 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:27:23,509 - model - INFO - model: ASLM Model - Waiting 4
2023-07-25 14:27:24,009 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:27:24,009 - model - INFO - model: ASLM Model - Waiting 3
2023-07-25 14:27:24,511 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:27:24,511 - model - INFO - model: ASLM Model - Waiting 2
2023-07-25 14:27:25,012 - model - INFO - model: ASLM Model - Running data process, get frames []
2023-07-25 14:27:25,012 - model - INFO - model: ASLM Model - Waiting 1
2023-07-25 14:27:25,012 - model - INFO - model: ASLM Model - Data thread stopped.
2023-07-25 14:27:25,013 - model - INFO - model: ASLM Model - Received frames in total: 1
```

zacsimile commented:
On reboot, this works fine.

AdvancedImagingUTSW committed:
Fix NoneType channel issue

Co-Authored-By: Zach Marin <zacsimile@users.noreply.github.com>

AdvancedImagingUTSW committed:
More attempts at stage stability

Co-Authored-By: Zach Marin <zacsimile@users.noreply.github.com>

annie-xd-wang committed:
small fix

annie-xd-wang committed:
Merge branch 'develop' into new-sutter-fixes

annie-xd-wang committed:
small fix
